File: server/src/modules/users/dto/user-profile.dto.ts
import type { AuthUserEntity } from '@app/modules/auth/entities/auth-user.entity';
import type { FriendshipRelationshipView } from '@app/modules/friendships/dto/friendship.dto';

export interface UserProfileDto {
  id: string;
  identifier: string;
  nickname: string;
  handle: string;
  avatarUrl: string | null;
  discoveryMode: 'public' | 'private';
}

export interface DiscoverableUserDto {
  discoverable: boolean;
  profile: UserDiscoveryProfileDto | null;
  relationship: FriendshipRelationshipView;
}

export interface UserDiscoveryProfileDto {
  id: string;
  nickname: string;
  handle: string;
  avatarUrl: string | null;
}

export function toUserProfileDto(user: AuthUserEntity): UserProfileDto {
  return {
    id: user.id,
    identifier: user.identifier,
    nickname: user.nickname,
    handle: user.handle,
    avatarUrl: user.avatarUrl,
    discoveryMode: user.discoveryMode,
  };
}

export function toUserDiscoveryProfileDto(
  user: AuthUserEntity,
): UserDiscoveryProfileDto {
  return {
    id: user.id,
    nickname: user.nickname,
    handle: user.handle,
    avatarUrl: user.avatarUrl,
  };
}

File: mobile/lib/features/profile/data/dto/user_profile_dto.dart
import 'package:production_chat_app/features/profile/domain/entities/user_profile.dart';

class UserProfileDto {
  const UserProfileDto({
    required this.id,
    required this.identifier,
    required this.nickname,
    required this.handle,
    required this.avatarUrl,
    required this.discoveryMode,
  });

  factory UserProfileDto.fromJson(Map<String, dynamic> json) {
    return UserProfileDto(
      id: json['id']?.toString() ?? '',
      identifier: json['identifier']?.toString() ?? '',
      nickname: json['nickname']?.toString() ?? '',
      handle: json['handle']?.toString() ?? '',
      avatarUrl: json['avatarUrl']?.toString(),
      discoveryMode: json['discoveryMode']?.toString() ?? 'public',
    );
  }

  final String id;
  final String identifier;
  final String nickname;
  final String handle;
  final String? avatarUrl;
  final String discoveryMode;

  UserProfile toEntity() {
    return UserProfile(
      id: id,
      identifier: identifier,
      nickname: nickname,
      handle: handle,
      avatarUrl: avatarUrl,
      discoveryMode: discoveryMode,
    );
  }
}
